# Q1

In [1]:
!mlflow --version

mlflow, version 1.26.0


# Q2

In [1]:
!python utils/preprocess_data.py --raw_data_path ~/Documents/GitHub/mlops-zoomcamp-my-repo/data/ --dest_path ./output

In [2]:
!ls output

dv.pkl    test.pkl  train.pkl valid.pkl


There are 4 files in the output folder.

# Q3

In [5]:
!python utils/train.py

2022/05/30 17:32:12 INFO mlflow.tracking.fluent: Experiment with name 'homework-week-2' does not exist. Creating a new experiment.
2022/05/30 17:32:12 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
2022/05/30 17:32:22 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/tongqiao/miniforge3/envs/mlops/lib/python3.9/site-packages/_distutils_hack/__init__.py:30: UserWarning: Setuptools is replacing distutils."


In [8]:
from mlflow.tracking import MlflowClient

MLFLOW_TRACKING_URI = "sqlite:///mlflow.db"
client = MlflowClient(tracking_uri=MLFLOW_TRACKING_URI)

In [30]:
run_id = "60ef7eab698842fc8a008fed2bc60c7c"
run = client.search_runs(experiment_ids="1")[0]
print("There are {} parameters logged by MLflow.".format(len(run.data.params)))

There are 17 parameters logged by MLflow.


# Q4

Command to launch the tracking server:
```
mlflow ui --backend-store-uri sqlite:///mlflow.db --artifacts-destination ./artifacts
```

# Q5

In [ ]:
!python utils/hpo.py

In [40]:
runs = client.search_runs(
    experiment_ids="2",
    filter_string="",
    max_results=1,
    order_by=["metrics.rmse ASC"]
)
runs
    

[<Run: data=<RunData: metrics={'rmse': 6.6284257482044735}, params={'max_depth': '19',
  'min_samples_leaf': '3',
  'min_samples_split': '5',
  'n_estimators': '28',
  'random_state': '42'}, tags={'mlflow.source.git.commit': '7bfb6bf2b36aa6428bf8ac89d5e500f74f18bf1b',
  'mlflow.source.name': 'utils/hpo.py',
  'mlflow.source.type': 'LOCAL',
  'mlflow.user': 'tongqiao',
  'model': 'random-forest-regressor'}>, info=<RunInfo: artifact_uri='./mlruns/2/746c826f6ca5432492d57f3af309e2ff/artifacts', end_time=1653930438096, experiment_id='2', lifecycle_stage='active', run_id='746c826f6ca5432492d57f3af309e2ff', run_uuid='746c826f6ca5432492d57f3af309e2ff', start_time=1653930432181, status='FINISHED', user_id='tongqiao'>>]

The best validation RMSE is 6.628 as shown above.

# Q6

In [ ]:
!python utils/register_model.py

In [43]:
run = client.search_runs(
    experiment_ids="3",
    filter_string="",
    max_results=1,
    order_by=["metrics.rmse ASC"]
)[0]
run

<Run: data=<RunData: metrics={'test_rmse': 6.548852722401935,
 'training_mae': 3.7776220690634528,
 'training_mse': 30.58526647604336,
 'training_r2_score': 0.771248028190807,
 'training_rmse': 5.530394784827152,
 'training_score': 0.771248028190807,
 'valid_rmse': 6.629893219637507}, params={'bootstrap': 'True',
 'ccp_alpha': '0.0',
 'criterion': 'squared_error',
 'max_depth': '20',
 'max_features': '1.0',
 'max_leaf_nodes': 'None',
 'max_samples': 'None',
 'min_impurity_decrease': '0.0',
 'min_samples_leaf': '2',
 'min_samples_split': '2',
 'min_weight_fraction_leaf': '0.0',
 'n_estimators': '25',
 'n_jobs': 'None',
 'oob_score': 'False',
 'random_state': '42',
 'verbose': '0',
 'warm_start': 'False'}, tags={'estimator_class': 'sklearn.ensemble._forest.RandomForestRegressor',
 'estimator_name': 'RandomForestRegressor',
 'mlflow.log-model.history': '[{"run_id": "6488313603024130a90ab398d7024d08", '
                             '"artifact_path": "model", "utc_time_created": '
         

The test RMSE of the best model is 6.55 as shown above.